### PR TITLE
8264268: Don't use oop types for derived pointers

### PR DESCRIPTION
**Pull request by stefank**
The JIT compiler embeds pointers to addresses within an object. These are called derived pointers. When the GC moves objects, these pointers need to be updated explicitly, because the GC only deals with the real oops of the objects (base pointer).

The code that deals with this uses oop* for the address containing the base pointer. This is fine, the address contains an oop. However, it also uses oop* for the interior pointer, even though the contents is not a valid oop.

This creates temporary oops that does not conform to the normal requirements for oops. For example, the lower three bits could be set. This makes it problematic to write stricter verification code.

I propose that we use intptr_t* instead of oop*, and only use oop* when the location is known to contain a valid oop.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8264268](https://bugs.openjdk.java.net/browse/JDK-8264268): Don't use oop types for derived pointers

### Reviewers
 * [John R Rose](https://openjdk.java.net/census#jrose) (@rose00 - **Reviewer**) ⚠️ Review applies to 38f494a4ba0ae7b52c27081eeedd0549ee849a93
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**) ⚠️ Review applies to d231e3baea33763387b09f6cad262ce009b2f1e6

### Download
To checkout this PR locally:
`$ git fetch https://git.openjdk.java.net/jdk pull/3214/head:pull/3214`
`$ git checkout pull/3214`

To update a local copy of the PR:
`$ git checkout pull/3214`
`$ git pull https://git.openjdk.java.net/jdk pull/3214/head`

### CODE REVIEW

**Review by rose00 — APPROVED** [review]
Looks good.

Maybe add `typedef intptr_t derived_oop_t` to make the operations on d-oops a little more distinctive.

Maybe add a static assert that `sizeof(oop) == sizeof(intptr_t)`, just in case there's a problem with compressed oops in the future.

**Review by kimbarrett — COMMENTED** [review]
I agree with jrose that having a type name for derived pointers would be better.  But I wondered if it might be possible to make it stronger than just an alias type for intptr_t.  I tried making it an enum class with a few supporting operations, and it doesn't look too bad.
https://github.com/kimbarrett/openjdk-jdk/tree/derived_ptr

**Comment by stefank** [comment]
> Looks good.

Thanks.

> 
> Maybe add `typedef intptr_t derived_oop_t` to make the operations on d-oops a little more distinctive.

I used Kim's enum suggestion.

> 
> Maybe add a static assert that `sizeof(oop) == sizeof(intptr_t)`, just in case there's a problem with compressed oops in the future.

Done.

**Comment by stefank** [comment]
> I agree with jrose that having a type name for derived pointers would be better. But I wondered if it might be possible to make it stronger than just an alias type for intptr_t. I tried making it an enum class with a few supporting operations, and it doesn't look too bad.
> https://github.com/kimbarrett/openjdk-jdk/tree/derived_ptr

Sounds good. I've taken your patch and and did some minor cleanups.

File: src/hotspot/share/compiler/oopMap.cpp [hunk 1]
```diff
@@ -284,10 +285,9 @@ void OopMapSet::all_do(const frame *fr, const RegisterMap *reg_map,
       }
 #endif
 #endif // !COMPILER2
-      oop* loc = fr->oopmapreg_to_location(omv.reg(),reg_map);
-      guarantee(loc != NULL, "missing saved register");
-      oop *derived_loc = loc;
-      oop *base_loc    = fr->oopmapreg_to_location(omv.content_reg(), reg_map);
+      intptr_t* derived_loc = (intptr_t*)fr->oopmapreg_to_location(omv.reg(),reg_map);
+      guarantee(derived_loc != NULL, "missing saved register");
+      oop *base_loc = fr->oopmapreg_to_oop_location(omv.content_reg(), reg_map);
```
[the remark this answers is not in the document]
Comment (stefank) [reply src/hotspot/share/compiler/oopMap.cpp]: Done

**Review by kimbarrett — APPROVED** [review]

File: src/hotspot/share/compiler/oopMap.cpp [hunk 2]
```diff
@@ -193,23 +193,24 @@ void OopMapSet::add_gc_map(int pc_offset, OopMap *map ) {
   add(map);
 }
 
-static void add_derived_oop(oop* base, oop* derived, OopClosure* oop_fn) {
+static void add_derived_oop(oop* base, intptr_t* derived, OopClosure* oop_fn) {
```
[the remark this answers is not in the document]
Comment (stefank) [reply src/hotspot/share/compiler/oopMap.cpp]: Hmm. `&do_nothing_cl` isn't being passed down to `oop_fn`. It's being passed down as `value_fn`, which isn't used at all. I've pushed a commit to remove this, to make it less confusing.

**Pull request closed without merging**